core/src/main/java/org/jdbi/v3/core/Handle.java

```diff
@@ -122,47 +124,57 @@ public Handle setStatementBuilder(StatementBuilder builder) {
      */
     @Override
     public void close() {
+        final List<Throwable> suppressed = new ArrayList<>();
         if (closed) {
             return;
         }
 
-        boolean wasInTransaction = isInTransaction()
-            && forceEndTransactions
-            && config.get().get(Handles.class).isForceEndTransactions();
+        boolean wasInTransaction = false;
+        if (forceEndTransactions && config.get().get(Handles.class).isForceEndTransactions()) {
+            try {
+                wasInTransaction = isInTransaction();
+            } catch (Exception e) {
+                suppressed.add(e);
+            }
+        }
 
         extensionMethod.remove();
         config.remove();
 
         if (wasInTransaction) {
-            rollback();
+            try {
+                rollback();
+            } catch (Exception e) {
+                suppressed.add(e);
+            }
         }
 
-        Exception suppressed = null;
         try {
             statementBuilder.close(getConnection());
         } catch (Exception e) {
-            suppressed = e;
+            suppressed.add(e);
         }
 
         try {
             closer.close(connection);
 
+            if (!suppressed.isEmpty()) {
+                final Throwable original = suppressed.remove(0);
+                suppressed.forEach(original::addSuppressed);
+                throw new CloseException("Failed to clear transaction status on close", original);
+            }
             if (wasInTransaction) {
                 TransactionException txe = new TransactionException("Improper transaction handling detected: A Handle with an open "
                     + "transaction was closed. Transactions must be explicitly committed or rolled back "
                     + "before closing the Handle. "
                     + "Jdbi has rolled back this transaction automatically. "
                     + "This check may be disabled by calling getConfig(Handles.class).setForceEndTransactions(false).");
-                if (suppressed != null) {
-                    txe.addSuppressed(suppressed);
-                }
+                suppressed.forEach(txe::addSuppressed);
```

Comment:
This lambda code is dead because `suppressed` list is **always** empty here. This is because for non-empty `supressed` list the `if` block in 161 is executed which throws `CloseException` in 164 making this one unreachable. Therefore, the only way for this line to run is to have `suppressed` list empty, but, because it's empty the lambda within `forEach` will never run.

I suggest to unify suppression propagation in a `catch` blocks to make it simpler:

```java
try {
  closer.close()
  if (wasInTransaction && suppressed.isEmpty()) {
    throw new TransactionException("Improper ...");
  }
} catch (Throwable t) {
  CloseException ce = new CloseException("Unable to close Connection", t);
  suppressed.forEach(ce::addSuppressed);
  throw ce;
} finally {
  closed = true;
}

if (!suppressed.isEmpty()) {
  CloseException ce = new CloseException("Failed to clear transaction status on close", suppressed.remove(0));
  suppressed.forEach(ce::addSuppressed);
  throw ce;  
}
```

The above code is just to present an idea. I was writing it without IDE so it may may not compile or may be buggy or violate some dependencies or requirements in call stack which I'm not aware of.

Reply:
I think @sarxos has it right. This line of code is unreachable if `suppressed` is non-empty. So this line is superfluous.